solver/src/liquidity/offchain_orderbook.rs

```diff
@@ -47,6 +51,194 @@ pub fn normalize_limit_order(order: Order, native_token: WETH9) -> LimitOrder {
 
 impl SettlementHandling<LimitOrder> for OrderSettlementHandler {
     fn encode(&self, executed_amount: U256, encoder: &mut SettlementEncoder) -> Result<()> {
-        encoder.add_trade(self.order.clone(), executed_amount)
+        if self.order.order_creation.buy_token == BUY_ETH_ADDRESS {
+            let interaction = compute_unwrap_interaction(
+                encoder.clearing_prices(),
+                &self.order,
+                executed_amount,
+                self.native_token.clone(),
+            )?;
+
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+
+            encoder.add_unwrap(interaction);
+        } else {
+            encoder.add_trade(self.order.clone(), executed_amount)?;
+        }
+
+        Ok(())
+    }
+}
+
+fn compute_unwrap_interaction(
+    clearing_prices: &HashMap<H160, U256>,
+    order: &Order,
+    executed_amount: U256,
+    weth: WETH9,
+) -> Result<UnwrapWethInteraction> {
+    let sell_price = if let Some(&price) = clearing_prices.get(&order.order_creation.sell_token) {
+        price
+    } else {
+        return Err(anyhow!("sell price not available"));
+    };
+    let buy_price = if let Some(&price) = clearing_prices.get(&order.order_creation.buy_token) {
+        price
+    } else {
+        return Err(anyhow!("buy price not available"));
+    };
+    let amount = if let Some(amount) = executed_buy_amount(
+        order,
+        executed_amount,
+        Price {
+            sell_price,
+            buy_price,
+        },
+    ) {
+        amount
+    } else {
+        return Err(anyhow!("cannot compute executed buy amount"));
+    };
+    Ok(UnwrapWethInteraction { weth, amount })
+}
+
+struct Price {
+    sell_price: U256,
+    buy_price: U256,
+}
+
+fn executed_buy_amount(order: &Order, executed_amount: U256, price: Price) -> Option<U256> {
+    Some(match order.order_creation.kind {
+        OrderKind::Sell => executed_amount
+            .checked_mul(price.sell_price)?
+            .checked_div(price.buy_price)?,
+        OrderKind::Buy => executed_amount,
+    })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::interactions::dummy_web3;
+    use crate::settlement::assert_settlement_encoded_with;
+    use maplit::hashmap;
+    use model::order::OrderCreation;
+
```

Comment:
Maybe add a unit test for `executed_buy_amount` on arithmetic overflow when computing executed amount?